### PR TITLE
PHP 8: Code Review `Membership`

### PR DESCRIPTION
**Pull request by playamongthestars**
Hello @smeyer-ilias, 

I completed the review of the `Services/Membership` component.

Results:

- [x] The code style is `PSR-2 + X` compliant
- [ ] No usage of `$_GET` / `$_POST` / `$_REQUEST` / `$_SESSION`
- [x] There is a `Unit Test Suite` with at least one unit test
- [x] The unit tests pass
- [x] External libraries are compatible with PHP 8 (if relevant)
- [x] There are no serious `PhpStorm Code Inspection` issues left
- [ ] The types are fully documented (PhpDoc) or explict PHP types are used (type hints, return types, typed properties)

--

Actions needed:
- [ ] Please check my inline comments. Some of them are just information, some require an action. You can use `TODO PHP8-REVIEW` when searching.

Best regards,
@nmatuschek

### CODE REVIEW

File: Services/Membership/classes/Export/class.ilMemberExport.php [hunk 1]
```diff
@@ -76,7 +76,7 @@ public function __construct(int $a_ref_id, int $a_type = self::EXPORT_CSV)
         $this->initGroups();
 
         $this->agreement = ilMemberAgreement::_readByObjId($this->obj_id);
-        $this->settings = new ilUserFormSettings('memexp');
+        $this->settings = new ilUserFormSettings('memexp'); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/Export/class.ilMemberExport.php:79]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/Export/class.ilMemberExportGUI.php [hunk 2]
```diff
@@ -87,7 +87,7 @@ public function executeCommand() : void
     protected function initSettingsForm(bool $a_is_excel = false) : ilPropertyFormGUI
     {
         // Check user selection
-        $this->exportSettings = new ilUserFormSettings('memexp');
+        $this->exportSettings = new ilUserFormSettings('memexp'); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/Export/class.ilMemberExportGUI.php:90]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/Export/class.ilMemberExportGUI.php [hunk 3]
```diff
@@ -228,7 +228,7 @@ protected function handleIncoming() : void
         }
 
         // Save (form) settings
-        $this->exportSettings = new ilUserFormSettings('memexp');
+        $this->exportSettings = new ilUserFormSettings('memexp'); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/Export/class.ilMemberExportGUI.php:231]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/Export/class.ilMemberExportSettingsGUI.php [hunk 4]
```diff
@@ -157,10 +170,10 @@ protected function initForm($a_type) : ilPropertyFormGUI
             $identifier_for_object = $identifier . '_0';
         }
 
-        $settings = new ilUserFormSettings($identifier_for_object, -1);
+        $settings = new ilUserFormSettings($identifier_for_object, -1); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/Export/class.ilMemberExportSettingsGUI.php:173]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/Export/class.ilMemberExportSettingsGUI.php [hunk 5]
```diff
@@ -157,10 +170,10 @@ protected function initForm($a_type) : ilPropertyFormGUI
...
         if (!$settings->hasStoredEntry()) {
             // use default settings
-            $settings = new ilUserFormSettings($identifier, -1);
+            $settings = new ilUserFormSettings($identifier, -1); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/Export/class.ilMemberExportSettingsGUI.php:176]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/Export/class.ilMemberExportSettingsGUI.php [hunk 6]
```diff
@@ -184,7 +197,7 @@ protected function savePrintViewSettings() : void
                 $identifier .= '_' . $this->parent_obj_id;
             }
 
-            $settings = new ilUserFormSettings($identifier, -1);
+            $settings = new ilUserFormSettings($identifier, -1); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/Export/class.ilMemberExportSettingsGUI.php:200]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/class.ilAttendanceList.php [hunk 7]
```diff
@@ -16,7 +16,7 @@ class ilAttendanceList
     protected ilObject $parent_obj;
     protected ?ilParticipants $participants;
     protected ?ilWaitingList $waiting_list;
-    protected $callback;
+    protected $callback;  // TODO PHP8-REVIEW: Missing property's type declaration
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilAttendanceList.php:19]: // TODO PHP8-REVIEW: Missing property's type declaration

File: Services/Membership/classes/class.ilAttendanceList.php [hunk 8]
```diff
@@ -357,19 +357,19 @@ public function initForm(string $a_cmd = "") : ilPropertyFormGUI
         $form->addCommandButton($a_cmd, $this->lng->txt('sess_print_attendance_list'));
 
         if ($this->id && $a_cmd) {
-            $settings = new ilUserFormSettings($this->id);
+            $settings = new ilUserFormSettings($this->id); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilAttendanceList.php:359]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/class.ilAttendanceList.php [hunk 9]
```diff
@@ -357,19 +357,19 @@ public function initForm(string $a_cmd = "") : ilPropertyFormGUI
...
             if (!$settings->hasStoredEntry()) {
-                $settings = new ilUserFormSettings($this->parent_obj->getType() . 's_pview', -1);
+                $settings = new ilUserFormSettings($this->parent_obj->getType() . 's_pview', -1); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilAttendanceList.php:361]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/class.ilAttendanceList.php [hunk 10]
```diff
@@ -357,19 +357,19 @@ public function initForm(string $a_cmd = "") : ilPropertyFormGUI
...
             }
 
             $settings->deleteValue('desc'); // #11340
             $settings->exportToForm($form);
         } elseif ($a_cmd == 'printForMembersOutput') {
             $settings = new ilUserFormSettings($this->parent_obj->getType() . 's_pview_' . $this->parent_obj->getId(),
-                -1);
+                -1); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilAttendanceList.php]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/class.ilAttendanceList.php [hunk 11]
```diff
@@ -357,19 +357,19 @@ public function initForm(string $a_cmd = "") : ilPropertyFormGUI
...
             if (!$settings->hasStoredEntry()) {
                 // init from global defaults
-                $settings = new ilUserFormSettings($this->parent_obj->getType() . 's_pview', -1);
+                $settings = new ilUserFormSettings($this->parent_obj->getType() . 's_pview', -1); // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilAttendanceList.php:373]: // TODO PHP8-REVIEW Expected parameter of type 'int', 'string' provided

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 12]
```diff
@@ -526,13 +526,13 @@ public function updateParticipants() : void
             // permission, make sure he doesn't remove the course
             // administrator role of members who are course administrator.
             if (!$hasEditPermissionAccess && $memberIsAdmin &&
-                !in_array($adminRoleId, $_POST['roles'][$usr_id])
+                !in_array($adminRoleId, $_POST['roles'][$usr_id])  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:529]: / TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 13]
```diff
@@ -526,13 +526,13 @@ public function updateParticipants() : void
...
             ) {
-                $_POST['roles'][$usr_id][] = $adminRoleId;
+                $_POST['roles'][$usr_id][] = $adminRoleId; // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:531]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 14]
```diff
@@ -526,13 +526,13 @@ public function updateParticipants() : void
...
             }
 
             // Validate the role ids in the post data
-            foreach ((array) $_POST['roles'][$usr_id] as $role_id) {
+            foreach ((array) $_POST['roles'][$usr_id] as $role_id) {  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:535]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 15]
```diff
@@ -548,17 +548,17 @@ public function updateParticipants() : void
 
         $has_admin = false;
         foreach ($this->getMembersObject()->getAdmins() as $admin_id) {
-            if (!isset($_POST['roles'][$admin_id])) {
+            if (!isset($_POST['roles'][$admin_id])) {  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:551]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 16]
```diff
@@ -548,17 +548,17 @@ public function updateParticipants() : void
...
                 $has_admin = true;
                 break;
             }
-            if (in_array($adminRoleId, $_POST['roles'][$admin_id])) {
+            if (in_array($adminRoleId, $_POST['roles'][$admin_id])) { // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:555]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 17]
```diff
@@ -548,17 +548,17 @@ public function updateParticipants() : void
...
                 $has_admin = true;
                 break;
             }
         }
 
-        if (!$has_admin && is_array($_POST['roles'])) {
+        if (!$has_admin && is_array($_POST['roles'])) { // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:561]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 18]
```diff
@@ -573,19 +573,19 @@ public function updateParticipants() : void
         }
 
         foreach ($participants as $usr_id) {
-            $this->getMembersObject()->updateRoleAssignments($usr_id, (array) $_POST['roles'][$usr_id]);
+            $this->getMembersObject()->updateRoleAssignments($usr_id, (array) $_POST['roles'][$usr_id]);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilMembershipGUI.php:576]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 19]
```diff
@@ -121,7 +121,7 @@ protected function saveFields() : void
     {
         $fields = ilCourseDefinedFieldDefinition::_getFields($this->getObjId());
         foreach ($fields as $field_obj) {
-            $field_obj->enableRequired(isset($_POST['required'][$field_obj->getId()]));
+            $field_obj->enableRequired(isset($_POST['required'][$field_obj->getId()]));  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:124]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 20]
```diff
@@ -218,13 +218,13 @@ protected function saveField() : void
 
     protected function editField() : void
     {
-        if (!$_REQUEST['field_id']) {
+        if (!$_REQUEST['field_id']) {  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:221]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 21]
```diff
@@ -218,13 +218,13 @@ protected function saveField() : void
...
             $this->listFields();
             return;
         }
 
         $this->initFieldForm(self::MODE_UPDATE);
-        $udf = new ilCourseDefinedFieldDefinition($this->getObjId(), (int) $_REQUEST['field_id']);
+        $udf = new ilCourseDefinedFieldDefinition($this->getObjId(), (int) $_REQUEST['field_id']);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:227]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 22]
```diff
@@ -237,7 +237,7 @@ protected function updateField() : void
     {
         $this->initFieldForm(self::MODE_UPDATE);
         if ($this->form->checkInput()) {
-            $udf = new ilCourseDefinedFieldDefinition($this->getObjId(), (int) $_REQUEST['field_id']);
+            $udf = new ilCourseDefinedFieldDefinition($this->getObjId(), (int) $_REQUEST['field_id']);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:240]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 23]
```diff
@@ -274,7 +274,7 @@ protected function initFieldForm($a_mode) : ilPropertyFormGUI
                 break;
 
             case self::MODE_UPDATE:
-                $this->ctrl->setParameter($this, 'field_id', (int) $_REQUEST['field_id']);
+                $this->ctrl->setParameter($this, 'field_id', (int) $_REQUEST['field_id']);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:277]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 24]
```diff
@@ -338,13 +338,13 @@ protected function editMember(?ilPropertyFormGUI $form = null) : void
             ilMemberAgreementGUI::setCourseDefinedFieldValues(
                 $form,
                 $this->getObjId(),
-                (int) $_REQUEST['member_id']
+                (int) $_REQUEST['member_id']  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:341]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 25]
```diff
@@ -354,7 +354,7 @@ protected function initMemberForm() : ilPropertyFormGUI
         $form = new ilPropertyFormGUI();
         $form->setFormAction($this->ctrl->getFormAction($this));
         $title = $this->lng->txt(ilObject::_lookupType($this->getObjId()) . '_cdf_edit_member');
-        $name = ilObjUser::_lookupName((int) $_REQUEST['member_id']);
+        $name = ilObjUser::_lookupName((int) $_REQUEST['member_id']);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:357]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 26]
```diff
@@ -376,12 +376,12 @@ protected function saveMember() : void
         $form = $this->initMemberForm();
         if ($form->checkInput()) {
             // save history
-            $history = new ilObjectCustomUserFieldHistory($this->getObjId(), (int) $_REQUEST['member_id']);
+            $history = new ilObjectCustomUserFieldHistory($this->getObjId(), (int) $_REQUEST['member_id']);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:379]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php [hunk 27]
```diff
@@ -376,12 +376,12 @@ protected function saveMember() : void
...
             $history->setEditingTime(new ilDateTime(time(), IL_CAL_UNIX));
             $history->setUpdateUser($this->user->getId());
             $history->save();
 
-            ilMemberAgreementGUI::saveCourseDefinedFields($form, $this->getObjId(), (int) $_REQUEST['member_id']);
+            ilMemberAgreementGUI::saveCourseDefinedFields($form, $this->getObjId(), (int) $_REQUEST['member_id']);  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilObjectCustomUserFieldsGUI.php:384]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilRegistrationGUI.php [hunk 28]
```diff
@@ -249,7 +262,7 @@ protected function showCustomFields() : void
             switch ($field_obj->getType()) {
                 case ilCourseDefinedFieldDefinition::IL_CDF_TYPE_SELECT:
                     $select = new ilSelectInputGUI($field_obj->getName(), 'cdf[' . $field_obj->getId() . ']');
-                    $select->setValue(ilUtil::stripSlashes($_POST['cdf'][$field_obj->getId()]));
+                    $select->setValue(ilUtil::stripSlashes($_POST['cdf'][$field_obj->getId()]));  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilRegistrationGUI.php:264]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/class.ilRegistrationGUI.php [hunk 29]
```diff
@@ -259,7 +272,7 @@ protected function showCustomFields() : void
 
                 case ilCourseDefinedFieldDefinition::IL_CDF_TYPE_TEXT:
                     $text = new ilTextInputGUI($field_obj->getName(), 'cdf[' . $field_obj->getId() . ']');
-                    $text->setValue(ilUtil::stripSlashes($_POST['cdf'][$field_obj->getId()]));
+                    $text->setValue(ilUtil::stripSlashes($_POST['cdf'][$field_obj->getId()]));  // TODO PHP8-REVIEW Please fix this
```
Comment (playamongthestars) [inline Services/Membership/classes/class.ilRegistrationGUI.php:274]: // TODO PHP8-REVIEW Please fix this

File: Services/Membership/classes/Export/class.ilMemberExportFileTableGUI.php [hunk 30]
```diff
@@ -45,7 +45,7 @@ public function numericOrdering(string $a_field) : bool
         return in_array($a_field, array('size', 'date'));
     }
 
-    protected function getFiles(ilFileSystemStorage $a_fss_export) : void
+    protected function getFiles(ilFileSystemStorage $a_fss_export) : void// TODO PHP8-REVIEW The class `ilFileSystemStorage` does not exist
```
Comment (mjansenDatabay) [inline Services/Membership/classes/Export/class.ilMemberExportFileTableGUI.php:48]: / TODO PHP8-REVIEW The class `ilFileSystemStorage` does not exist

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 31]
```diff
@@ -1567,14 +1567,14 @@ protected function initAttendanceList(bool $a_for_members = false) : ?ilAttendan
             $part = ilParticipants::getInstance($member_id);
 
             $list = new ilAttendanceList(
-                $this,
+                $this,// TODO PHP8-REVIEW Argument does not match parameter type
```
Comment (mjansenDatabay) [inline Services/Membership/classes/class.ilMembershipGUI.php:1570]: TODO PHP8-REVIEW Argument does not match parameter type

File: Services/Membership/classes/class.ilMembershipGUI.php [hunk 32]
```diff
@@ -1567,14 +1567,14 @@ protected function initAttendanceList(bool $a_for_members = false) : ?ilAttendan
...
                 $this->getParentObject(),
                 $part,
                 $waiting_list
             );
         } else {
             $list = new ilAttendanceList(
-                $this,
+                $this,// TODO PHP8-REVIEW Argument does not match parameter type
```
Comment (mjansenDatabay) [inline Services/Membership/classes/class.ilMembershipGUI.php:1577]: TODO PHP8-REVIEW Argument does not match parameter type

File: Services/Membership/classes/Export/class.ilMemberExportGUI.php (hunk 2 above)
Comment (smeyer-ilias) [reply Services/Membership/classes/Export/class.ilMemberExportGUI.php:90]: This and all other calls of ilUserFormSettings were a wrong refactoring in Services/User. I fixed this in the trunk.
Now the expected first parameter is "string"

**Review by smeyer-ilias — APPROVED** [review]
Thanks for reviewing and fixing many issues like casting db values and using the refinery for not replaced GET, POST variables.

I will fix the remaining issues directly in the trunk.

**Pull request merged**